### PR TITLE
Logging - Log MoE token metrics for all kinds of aux losses

### PR DESCRIPTION
**Pull request by THUzxj**
Add tokens routed metrics for all kinds of MoE aux losses, and make such logging optional.

### CODE REVIEW

File: megatron/core/transformer/moe/router.py [hunk 1]
```diff
@@ -187,6 +187,15 @@ def _sinkhorn_activation(logits):
             _, indices = torch.topk(logits, k=self.topk, dim=1)
         map = torch.zeros_like(logits).int().scatter(1, indices, 1).bool()
         scores = logits * map
+        if self.training:
```
Comment (yzygitzh) [inline megatron/core/transformer/moe/router.py]: Please align the spacing before and after the code block to save metrics as in `global_batch_loss_load_balancing`.

File: megatron/core/transformer/moe/router.py [hunk 2]
```diff
@@ -187,6 +187,15 @@ def _sinkhorn_activation(logits):
...
+            tokens_per_expert = map.sum(dim=0)
+
+            save_to_tokens_per_expert_tracker(
+                "sinkhorn_tokens_per_expert",
```
Comment (yukirora) [inline megatron/core/transformer/moe/router.py:195]: please as argument to enable or disable the metrics

Comment (THUzxj) [reply megatron/core/transformer/moe/router.py:195]: As each training can use only one router type, I add one argument `log_tokens_per_expert` to control whether to log.

File: megatron/training/arguments.py [hunk 3]
```diff
@@ -2589,6 +2589,8 @@ def _add_moe_args(parser):
                        help='Add noise to the input tensor by applying jitter with a specified epsilon value.')
     group.add_argument('--moe-per-layer-logging', action='store_true',
                        help='Enable per-layer logging for MoE, currently supports auxiliary loss and z loss.')
+    group.add_argument('--moe-tokens-per-expert-logging', action='store_true',
```
Comment (yzygitzh) [inline megatron/training/arguments.py]: Shall we change to `--moe-tokens-logging`? The metrics (tokens, etc.) and the granularity (per-layer, per-expert) should be separated. The granularity should be controlled by `--moe-per-layer-logging` and in the future `--moe-per-expert-logging`.

File: megatron/core/transformer/moe/router.py [hunk 4]
```diff
@@ -39,6 +39,7 @@ def __init__(self, config: TransformerConfig) -> None:
         self.num_experts = self.config.num_moe_experts
         self.moe_aux_loss_func = None
         self.layer_number = None
+        self.moe_tokens_per_expert_logging = self.config.moe_tokens_per_expert_logging
```
Comment (yzygitzh) [inline megatron/core/transformer/moe/router.py]: Shall we just use `self.config.moe_tokens_logging` in the following and remove this line? It's actually more aligned with existing code.

**Comment by THUzxj** [comment]
Should I also use `moe_tokens_logging` to control the enable of tokens logging in global batch loss, which is already enabled in default? 

```
            save_to_tokens_per_expert_tracker(
                "global_batch_tokens_per_expert",
                tokens_per_expert,
                self.layer_number,
                self.config.num_layers,
                reduce_group=parallel_state.get_data_parallel_group())
```

**Comment by yzygitzh** [comment]
> Should I also use `moe_tokens_logging` to control the enable of tokens logging in global batch loss, which is already enabled in default?
> 
> ```
>             save_to_tokens_per_expert_tracker(
>                 "global_batch_tokens_per_expert",
>                 tokens_per_expert,
>                 self.layer_number,
>                 self.config.num_layers,
>                 reduce_group=parallel_state.get_data_parallel_group())
> ```

Yes all should be optional.

**Comment by THUzxj** [comment]
> > Should I also use `moe_tokens_logging` to control the enable of tokens logging in global batch loss, which is already enabled in default?
> > ```
> >             save_to_tokens_per_expert_tracker(
> >                 "global_batch_tokens_per_expert",
> >                 tokens_per_expert,
> >                 self.layer_number,
> >                 self.config.num_layers,
> >                 reduce_group=parallel_state.get_data_parallel_group())
> > ```
> 
> Yes all should be optional.

Thanks. This feature has been added.

**Review by yukirora — APPROVED** [review]

**Review by yzygitzh — APPROVED** [review]
